### ISSUE TITLE
Allowing a lint fails on let binding.

### ISSUE THREAD

**Issue by vi** [rust-lang/rust#51942]
Example with `cost_err`:

```
fn main() {
    #[allow(const_err)]
    let _x = 42 / 0;
}
```

```
#[allow(const_err)]
fn main() {
    let _x = 42 / 0;
}
```

Expected: the same. Actual: warning in the first case.

**Comment by oli-obk** [comment]
This is not just a `const_err` thing. This issue exists for all lints.

**Comment by ehuss** [comment]
I believe this is fixed via #69185 (nightly-2020-02-21).  `allow(unconditional_panic)` works on `let _x = 42 / 0;`.  And more convoluted lints also work (tested up to nightly-2021-05-12):

```rust
#[allow(const_err, unconditional_panic)]
let _x = [1; {const CONST: usize = 1/0; 1}];
```

I'm not sure what lints the previous comment is referring to. In my tests, other lint controls seem to work (back to 1.16, 1.15 seemed to ignore them).  I'll close, but if there is something else, please reopen with more information on what is not working.
